docs/en/pact-reference.md

```diff
@@ -1277,9 +1277,200 @@ Declaring models shares the same syntax with modules:
+directly-referenced code in the interests of code stability.
+
+### Important concerns when using modrefs.
+
+#### Late Binding
+
+As of Pact 4.7, modrefs are "late-binding", which means that the latest
+upgraded version of a module will be used when a module operation is invoked.
+
+Consider a modref to a module stored in the database when the module is
+at version 1. Sometime later the module is upgraded to version 2. The modref
+in the database will refer to the upgraded version 2 of the module when read
+back in and used.
+
+As described in the [Dependency Management](#dependency-management) section,
+Pact direct references are not late-binding, so this modref behavior might
+be surprising. "Pinned modrefs" that bind eagerly to the module version is
+being actively considered as an upcoming Pact change to extend the benefits of
+dependency management to modrefs.
+
+#### Modrefs introduce untrusted code
+
+As modrefs introduce the ability to call code unknown at compile time, this
+means that **any modref call should be treated as untrusted code**. Thus
+programmers should avoid invoking modrefs when a sensitive capability is in
+scope, as it will make "private capability functions" callable by the modref,
+when normally this would not be possible.
+
+Consider a module with a public function `collect-data` that is safe to call,
+and a private function `pay-fee` that is not intended to be invoked outside the
+module, guarded by the capability `COLLECT`. If a modref is invoked while the
+`COLLECT` capability is in scope, the calling code will be able to call the
+sensitive function directly.
+
+```lisp
+;; BAD CODE -- don't do this!
+(defun collect-data (collector:module{data-collector} account:string)
+  (with-capability (COLLECT)
+    ;; BAD: modref invoked inside of with-capability call
+    (store-data (collector::collect))
+    (pay-fee account)))
+
+(defun pay-fee (account:string)
+  (require-capability (COLLECT))
+  (coin.transfer FEE_BANK account FEE))
+```
+
+In the above code, `pay-fee` is intended to run once. But because the
+`collector:module{data-collector}` modref is invoked **inside of the capability
+acquisition**, the modref code in `collect` now has elevated privilege to
+directly call `pay-fee`. Thus a malicious actor can craft a version of `collect`
+thatn can directly call `pay-fee` as many times as they like.
+
+Fortunately, this is easily avoided by keeping modref calls out of scope of the
+sensitive capability.
+
+```lisp
+;; SAFE CODE -- do this!
+(defun collect-data (collector:module{data-collector} account:string)
+  ;; GOOD: modref invoked outside of with-capability call
+  (let ((data (collector::collect)))
+    (with-capability (COLLECT)
+      (store-data data)
+      (pay-fee account))))
+```
+
+Now, the modref call has safely returned before the capability is acquired,
+meaning a malicious implementation has no elevated privileges allowing it
+to call sensitive code.
```

Comment:
I'm really happy to see this included as part of the new documentation. Thanks for including it with a clear example!